### PR TITLE
Fix several index retention problems

### PR DESCRIPTION
**Pull request by bernd**
## TODO

- [x] Let `Deflector.cycle()` wait until the new index has been recovered and actually allocated.
- [ ] Let `Deflector.setUp()` wait until indices have been recovered.
- [ ] Let `IndexRotationTrhread.checkAndRepair()` wait until all indices have been recovered to avoid selecting a random index as new deflector.
- [ ] Let `Cluster.isConnectedAndHealthy` wait until the deflector index is ready by checking the index recovery status.
- [ ] Add an `end` field to the index ranges data which holds the timestamp of the newest message for an index. This is needed to make sure an index is included in a search if it contains messages with an unexpected timestamp. Graylog2/graylog2-web-interface#1238 #1200 
- [ ] Index ranges in MongoDB should not use "destroy all & re-create" in retention, but update/delete the range object in MongoDB for each index that is processed during retention. This leads to an empty index range collection during retention. #1167

## Notes

* Get list of indices via cluster state monitor or recovery API.
* The search should not try to use indices which are in recovery.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/indexer/indices/Indices.java [hunk 1]
```diff
@@ -394,4 +398,17 @@ public void optimizeIndex(String index) {
...
         c.admin().indices().optimize(or).actionGet();
     }
+
+    public void waitForRecovery(String index) {
+        RecoveryResponse response;
+
+        do {
+            LOG.warn("Waiting for index recovery for index {}", index);
+
+            final ActionFuture<RecoveryResponse> request = c.admin().indices().recoveries(new RecoveryRequest(index));
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/indexer/indices/Indices.java]: I think a `ClusterHealthRequest` for `index` with `waitForStatus` YELLOW or GREEN might be more useful here (https://www.elastic.co/guide/en/elasticsearch/reference/1.5/cluster-health.html#request-params) and doesn't require us to poll ourselves.

Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/indexer/indices/Indices.java]: Thanks for the hint!

Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/indexer/indices/Indices.java]: very good point!

**Comment by joschi** [comment]
Merged into `master`.

**Pull request closed without merging**